test/suites/upgrade.sh

```diff
@@ -0,0 +1,225 @@
+#!/bin/bash
+
+test_upgrade() {
+  reset_systems 4 2 3
+
+  # Perform upgrade test from MicroCloud 1 to 2.
+  if [ "${MICROCLOUD_SNAP_CHANNEL}" = "1/stable" ]; then
+    microceph_target="squid/edge" # TODO: squid/stable when available
+    microovn_target="latest/edge" # TODO: 24.03/stable when available
+    lxd_target="5.21/edge"
+    microcloud_target="latest/edge" # TODO: 2/stable when available
+
+    # The lookup subnet has to contain the netmask and the address has to be the one used by MicroCloud not the gateway.
+    lookup_subnet="$(lxc ls micro01 -f json -c4 | jq -r '.[0].state.network.enp5s0.addresses[] | select(.family == "inet") | .address + "/" + .netmask')"
+
+    # Use the MicroCloud version 1 preseed format with all the features available at the time of release:
+    # - local and remote storage
+    # - wiping disks
+    # - OVN uplink interface
+    preseed="$(cat << EOF
+lookup_subnet: ${lookup_subnet}
+systems:
+- name: micro01
+  ovn_uplink_interface: enp6s0
+  storage:
+    local:
+      path: /dev/disk/by-id/scsi-0QEMU_QEMU_HARDDISK_lxd_disk1
+      wipe: true
+    ceph:
+      - path: /dev/disk/by-id/scsi-0QEMU_QEMU_HARDDISK_lxd_disk2
+        wipe: true
+- name: micro02
+  ovn_uplink_interface: enp6s0
+  storage:
+    local:
+      path: /dev/disk/by-id/scsi-0QEMU_QEMU_HARDDISK_lxd_disk1
+      wipe: true
+    ceph:
+      - path: /dev/disk/by-id/scsi-0QEMU_QEMU_HARDDISK_lxd_disk2
+        wipe: true
+- name: micro03
+  ovn_uplink_interface: enp6s0
+  storage:
+    local:
+      path: /dev/disk/by-id/scsi-0QEMU_QEMU_HARDDISK_lxd_disk1
+      wipe: true
+    ceph:
+      - path: /dev/disk/by-id/scsi-0QEMU_QEMU_HARDDISK_lxd_disk2
+        wipe: true
+ovn:
+  ipv4_gateway: 10.1.123.1/24
+  ipv4_range: 10.1.123.100-10.1.123.254
+  ipv6_gateway: fd42:1:1234:1234::1/64
+EOF
+    )"
+
+    # Deactivate the fourth MicroCloud so we can add it later after upgrade.
+    lxc exec micro04 -- snap stop microcloud
+
+    # Deploy a version 1 MicroCloud and launch some instances.
+    lxc exec micro01 --env TEST_CONSOLE=0 -- microcloud init --preseed <<< "$preseed"
+
+    if [ "${SKIP_VM_LAUNCH}" = "1" ]; then
+      echo "::warning::SKIPPING VM LAUNCH TEST"
+    else
+      lxc exec micro01 -- lxc launch ubuntu:noble v1 --vm
```

Comment:
Should we use the same minimal-daily image we use elsewhere? Possibly with `security.devlxd.images=true`?

Reply:
Even using the `-daily` image isn't a silver bullet against needing to run `apt-get update`, unfortunately.

Reply:
I am not against using the minimal daily instead. I would not duplicate the image on the runner to prevent confusion later on when you expect to get a `ubuntu-minimal` but instead you get the `-daily`.

Reply:
I was just suggesting a possible cache seed. This wouldn't have any visible side effect as the requested image would be the same. The only difference would be that child VMs would download it much faster if it's coming from their parent's cache rather than from the network.

`security.devlxd.images=true` just tells LXD to check over the devlxd communication channel if the precise fingerprint it wants is downloadable over that fast devlxd connection or not. If yes, it downloads that image from there, if not, it uses the slower network to get what it needs.

Reply:
Yeah I did some tests before with `security.devlxd.images=true` but probably I did a mistake elsewhere as it didn't download the image from the host.

I'll shortly push another change that uses the `ubuntu-minimal-daily` also for the workload tests. In addition the equivalent container image gets pre fetched too so that the nested LXDs can consume it.

Thanks for the hint, wasn't aware of the feature till now :)

Reply:
Fixed.

As `$BASE_OS` can be different for the `micro*` VMs, there is now a step to pre fetch Ubuntu 22.04 and 24.04 images for both VMs and containers so we always have them at hand.